src/jdk.incubator.concurrent/share/classes/jdk/incubator/concurrent/ScopedValue.java

```diff
@@ -0,0 +1,839 @@
+     *     // @link substring="run" target="Carrier#run(Runnable)" :
+     *     ScopedValue.where(key, value).run(op);
+     * }
+     *
+     * @param key the ScopedValue
+     * @param value the value, can be {@code null}
+     * @param <T> the type of the value
+     * @param op the operation to call
+     */
+    public static <T> void where(ScopedValue<T> key, T value, Runnable op) {
+        where(key, value).run(op);
+    }
+
+    private ScopedValue() {
+        this.hash = generateKey();
+    }
+
+    /**
+     * Creates a scoped value that is initially unbound for all threads.
+     *
+     * @param <T> the type of the value
+     * @return a new {@code ScopedValue}
+     */
+    public static <T> ScopedValue<T> newInstance() {
+        return new ScopedValue<T>();
+    }
+
+    /**
+     * {@return the value of the scoped value if bound in the current thread}
+     *
+     * @throws NoSuchElementException if the scoped value is not bound
+     */
+    @ForceInline
+    @SuppressWarnings("unchecked")
+    public T get() {
+        Object[] objects;
+        if ((objects = scopedValueCache()) != null) {
+            // This code should perhaps be in class Cache. We do it
+            // here because the generated code is small and fast and
+            // we really want it to be inlined in the caller.
+            int n = (hash & Cache.SLOT_MASK) * 2;
+            if (objects[n] == this) {
+                return (T)objects[n + 1];
+            }
+            n = ((hash >>> Cache.INDEX_BITS) & Cache.SLOT_MASK) * 2;
+            if (objects[n] == this) {
+                return (T)objects[n + 1];
+            }
+        }
+        return slowGet();
+    }
+
+    @SuppressWarnings("unchecked")
+    private T slowGet() {
+        var value = findBinding();
+        if (value == Snapshot.NIL) {
+            throw new NoSuchElementException();
+        }
+        Cache.put(this, value);
+        return (T)value;
+    }
+
+    /**
+     * {@return {@code true} if this scoped value is bound in the current thread}
+     */
+    public boolean isBound() {
+        // ??? Do we want to search cache for this? In most cases we don't expect
+        // this {@link ScopedValue} to be bound, so it's not worth it. But I may
+        // be wrong about that.
+/*
+        if (Cache.find(this) != Snapshot.NIL) {
+            return true;
+        }
+ */
+        return findBinding() != Snapshot.NIL;
+    }
+
+    /**
+     * Return the value of the scoped value or NIL if not bound.
+     */
+    private Object findBinding() {
+        Object value = scopedValueBindings().find(this);
+        return value;
+    }
+
+    /**
+     * Returns the value of this scoped value if bound in the current thread, otherwise
+     * returns {@code other}.
+     *
+     * @param other the value to return if not bound, can be {@code null}
+     * @return the value of the scoped value if bound, otherwise {@code other}
+     */
+    public T orElse(T other) {
```

Comment:
From an API perspective, wouldn't return `Optional` a more consistent choice? `Optional` has all methods we need to deal with this kind of stuff...

Reply:
This comment is on orElse but I suspect you are suggesting that get() be changed to return Optional. I think we'll need to get more feedback/usage of this API before re-visiting that.

Reply:
Yes, my comment was really on `get` - that said, I note that saying get().get() would look odd (but maybe finding some other name for ScopedValue::get, such as find might work)

Reply:
It certainly would look odd. This API is, by design, as lightweight as it possibly can be, both from an implementation and a user's point of view. It's also intended to be as close as possible to an "invisible" parameter passed to all callees. From that point of view, `get()` is a wart. `get().get()` is just...

Reply:
IMHO there are ways to have the cake and eat it too. That is, we could have a couple of overloads:

```
T get() { ... } // throws NSME if not found
Optional<T> find() // returns empty optional if not found
```

Then, for simple use cases, code will stay the same as today. But, if users want to deal with optionality explicitly, they can call `find` and then call `orElse`, `map` or whatever they like.

Reply:
We expect isBound() will be used a lot and I think that is clearer (and cheaper) than find().isEmpty(). 

Time will tell on orElse/orElseThrow and whether they should be replaced with an Optional view. That is, I think your comments mostly apply to those two methods rather than get/isBound.

Reply:
Note that `ScopedValue` can currently be bound to `null`, but by using `Optional`, there would be no way to differentiate an unbound `ScopedValue` from one bound to `null`.

Reply:
> Note that `ScopedValue` can currently be bound to `null`, but by using `Optional`, there would be no way to differentiate an unbound `ScopedValue` from one bound to `null`.

That's right, an Optional view would have to deal with that.

Reply:
OK, so unless anyone has any /strong/ objections I'll mark this as Resolved.